target_postgres/sinks.py

```diff
@@ -2,9 +2,182 @@
+
+    def __init__(self, *args, **kwargs):
+        self.temp_table_name = self.generate_temp_table_name()
+        super().__init__(*args, **kwargs)
+    
+    def process_batch(self, context: dict) -> None:
+        """Process a batch with the given batch context.
+
+        Writes a batch to the SQL target. Developers may override this method
+        in order to provide a more efficient upload/upsert process.
+
+        Args:
+            context: Stream partition or context dictionary.
+        """
+        #First we need to be sure the main table is already created
+        self.connector.prepare_table(
+            full_table_name=self.full_table_name,
+            schema=self.schema,
+            primary_keys=self.key_properties,
+            as_temp_table=False,
+        )
+        #Create a temp table (Creates from the table above)
+        self.connector.create_temp_table_from_table(
+                from_table_name=self.full_table_name,
+                temp_table_name=self.temp_table_name
+        )
+        #Insert into temp table
+        self.bulk_insert_records(
+            full_table_name=self.temp_table_name,
+            schema=self.schema,
+            records=context["records"],
+        )
+        #Merge data from Temp table to main table
+        self.merge_upsert_from_table(
+            from_table_name=self.temp_table_name,
+            to_table_name=self.full_table_name,
+            schema=self.schema,
+            join_keys=self.key_properties
+        )
+        self.connector.truncate_table(self.temp_table_name)
+    
+    #Copied purely to help with type hints
+    @property
+    def connector(self) -> PostgresConnector:
+        """The connector object.
+
+        Returns:
+            The connector object.
+        """
+        return self._connector
+
+    def generate_temp_table_name(self):
+        return f"temp_{str(uuid.uuid4()).replace('-','_')}"
+    
+    def merge_upsert_from_table(
+            self, from_table_name: str, to_table_name: str, schema: dict, join_keys: List[str]
+    ) -> Optional[int]:
+        """Merge upsert data from one table to another.
+
+        Args:
+            target_table_name: The destination table name.
+            from_table_name: The source table name.
+            join_keys: The merge upsert keys, or `None` to append.
+
+        Return:
+            The number of records copied, if detectable, or `None` if the API does not
+            report number of records affected/inserted.
+
+        Raises:
+            NotImplementedError: if the merge upsert capability does not exist or is
+                undefined.
+        """
+        #TODO think about sql injeciton, issue here https://github.com/MeltanoLabs/target-postgres/issues/22
+
+
+        #INSERT
+        join_condition = " and ".join([f"temp.{key} = target.{key}" for key in join_keys])
+        where_condition = " and ".join([f"target.{key} is null" for key in join_keys])
+
+        insert_sql = f"""
```

Comment:
Issue #22  maybe we can use a SQL Alchemy construct here and avoid the SQL Injection stuff